Design to support ByRefLike types in Generics

AaronRobinsonMSFT opened the pull request:
Adding in a proposal to support ByRefLike type in Generics. This work is supported by https://github.com/dotnet/runtime/issues/63768 and contributes toward https://github.com/dotnet/runtime/issues/65112.

/cc @jkotas @davidwrighton @jaredpar @cston @RikkiGibson @AlekseyTs

AaronRobinsonMSFT committed:
Design to support ByRefLike types in Generics

AaronRobinsonMSFT committed:
Trim whitespace

davidwrighton commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,75 @@
+throw
+```
+
+When boxing due to a constrained call that cannot be made, instead of allocating a normal boxed object, an object of `InvalidBoxedObject` type will be created. It will have implementations of the various overridable object methods which throw `InvalidProgramException`, and interface dispatch shall have a special case for attempting to invoke an interface method on such an object, that will also t ...
````
What about the behavior of default interface methods.

davidwrighton replied:
Oh I see. I shouldn't read documents, bottom up.

davidwrighton commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
```
For these scenarios, please describe the new behavior of the IL opcode. Should it throw? etc.

cston commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
+    public class GenericParameterSupportsAnyTypeAttribute : Attribute
```
`sealed`?

cston commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+        /// <summary>
+        /// Generic Parameter Index.
+        /// </summary>
+        public int ParameterIndex { get; set; }
```
Should `ParameterIndex` be readonly?

cston commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,75 @@
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
````
`| AttributeTargets.Struct | AttributeTargets.Delegate | AttributeTargets.Interface`

teo-tsirpanis replied:
According to the proposal in dotnet/csharplang#5817 generic delegates would automatically support byref-like types. Why would applying the attribute on delegates still be needed?

AaronRobinsonMSFT replied:
> generic delegates would automatically support byref-like types

I don't see that in the proposal nor would that be possible at present. There is no way to support ByRefLike types in Generics in any form at present. Perhaps I am misunderstanding your point though.

teo-tsirpanis replied:
Oops, I misclicked. The proposal is dotnet/csharplang#5689 and the relevant point is https://github.com/dotnet/csharplang/pull/5689/files#diff-67b62b3a3e0ed8896aa17e6a455ddc5ca90586364864c7bfd280223999a98b3fR191.

AaronRobinsonMSFT replied:
> The proposal is https://github.com/dotnet/csharplang/pull/5689

Thanks.

> Why would applying the attribute on delegates still be needed?

Because the compiler still needs a way to convey that message and encode it. At present there is no way to encode that case and this attribute is an option. However, others exist - see various statements in this design doc by @jkotas and @MichalStrehovsky.

AaronRobinsonMSFT replied:
Moving indication to metadata flags.

cston commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; This IL sequence must resolve to a method implemented on `object`, or a default interface method.
+- Use of any Generic which doesn’t expect a ByRefLike parameter as a Generic parameter.
```
Should `isinst` and `unbox.any` be included?

AaronRobinsonMSFT replied:
I don't think we can get to the `unbox.any` instruction since we can't box. Not saying I shouldn't call out behavior, but do want to make sure I'm understanding the why it should be included.

For `isinst`, that is a good point. I _think_ you're right. Let me page back in how this instruction works so I can define the "why".

cston replied:
For `unbox.any`, is the following possible?
```
.method static !!T F<T>(object obj)
{
    ldarg.0
    unbox.any !!T
    ret
}
```

AaronRobinsonMSFT replied:
I think this will simply always fail because `obj` will never be able to be a `T` that is ByRefLike.

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+
+## Proposal
+
+A new Attribute API will be defined to indicate which Generic parameters are permissible to be of any type&mdash;including ByRefLike types. The Attribute could be used by the compiler to implement a generic non-constraint.
```
We have `GenericParameterAttributes` flags in metadata. Would it make more sense to use a bit in these flags for this?

AaronRobinsonMSFT replied:
Yes. I agree this is probably the better place for this support.

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `throw` &ndash; Requires an object reference on the stack so not directly impacted since boxing is not permitted.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
```
Does this mean that the generic type can have static field like that, but all attempts to access it via IL instructions or via reflection are going to fail?

jkotas replied:
Does it mean that this is going to be C# compile error?
```
class G<AnyT> where AnyT : any
{
    static AnyT s;
}
```

AaronRobinsonMSFT replied:
> Does it mean that this is going to be C# compile error?

I don't believe so. This would only occur if `AnyT` was a ByRefLike type _and_ was accessed at run-time. Is the point here that `static` fields are automatically initialized during type load?

RikkiGibson replied:
It seems like it would be best to make it an error in C#.

AaronRobinsonMSFT replied:
@RikkiGibson That is fine. My point was that the runtime will handle this during type instantiation. As far as what C# does that is up to the language, the runtime will handle the case either way.

jkotas commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,75 @@
+}
+```
+
+A new API will be implemented as a JIT intrinsic for determining if a parameter is ByRefLike. This API would represent a check to occur at JIT time code-gen to avoid taking paths that would be invalid for some values of `T`.
````
What happens if the path that is invalid for some some values of `T` is not gated by this API?

AaronRobinsonMSFT replied:
If one of the invalid instructions is used, an `InvalidProgramException` would be thrown. That is at least my current thinking. I need to go through the instructions again and ensure there aren't cases where we would instead need a "fail fast".

jkotas commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,75 @@
+```
+
+A new API will be implemented as a JIT intrinsic for determining if a parameter is ByRefLike. This API would represent a check to occur at JIT time code-gen to avoid taking paths that would be invalid for some values of `T`.
+
````
Is the C# compiler going to generate calls to this API in some cases, or is this API meant just for the user code to use?

AaronRobinsonMSFT replied:
The example case here is `ToArray()` on `Span<T>`. The implementation would do the check and during JIT time the branch would be all that was present in the code - so user code. I could also see the compiler using this for scenarios too. If that were the case though I assume you would prefer to see this under `System.Runtime.CompilerServices`?

MichalStrehovsky commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,75 @@
+throw
+```
+
+When boxing due to a constrained call that cannot be made, instead of allocating a normal boxed object, an object of `InvalidBoxedObject` type will be created. It will have implementations of the various overridable object methods which throw `InvalidProgramException`, and interface dispatch shall have a special case for attempting to invoke an interface method on such an object, that will also t ...
````
Why do we need the `InvalidBoxedObject`? Can't we just throw `InvalidProgram` during the boxing attempt?

MichalStrehovsky replied:
I think I see:

1. If we throw during boxing, we would throw "Type 'SomeRefLikeType' needs to be boxed to perform this call, but the type is byref-like." (not giving context on the fact what interface/base virtual call we try to make).
2. If we make an InvalidBoxedObject, we would throw "Cannot call method IFoo.SomeMethod on this type because it's byref-like" (not giving context on the type of `this` instead).

1 has less implementation complexity (it may actually already work, modulo the more specific message). 2 gives more detail in a way but also less detail in another way.

AaronRobinsonMSFT replied:
I was given some additional feedback on this approach. I'm going to simply apply (1) unless we determine something like (2) is needed.

MichalStrehovsky commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,75 @@
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
````
Why not `[AttributeUsage(AttributeTargets.GenericParameter)]` and then we don't need to deal with the index separately?

(But if we expect runtime to be reading this, Jan's suggestion to use the from GenericParamAttributes is even better - though it's a file format change. If the runtime won't be reading this, I would be more inclined towards an attribute.)

AaronRobinsonMSFT replied:
I agree. I am incline to use `GenericParamAttributes`. I'll be updating this part.

MichalStrehovsky commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
```
Will this have a type loader impact too or it's more about just relaxing validation in the type loader and defining behaviors of instructions when T is a byref-like type?

It kind of feels like we could also handle this in the runtime the same way we handle the existing constraints (`struct`, `new()`, etc.). But maybe validating this can be purely a tooling concern, not runtime concern, since we will have semantics for when IL operations don't make sense for a specific T.

```csharp
class Foo<AnyT> where AnyT : any
{
    // Do we get a TypeLoadException for List<AnyT> when JIT compiling this method
    // or do we get an InvalidProgramException when the List<T> constructor is trying
    // to get an empty array of T for the backing store?
    static object Get() => new List<AnyT>();
}
```

AaronRobinsonMSFT replied:
> Will this have a type loader impact too or it's more about just relaxing validation in the type loader and defining behaviors of instructions when T is a byref-like type?

The latter. My thought here was to push failure to JIT time. In the scenario above, if `Get()` was never called then it should be okay. Implementers could use the `IsByRefLike<T>()` to tailor the failure case if desired though.

jkotas replied:
```
class Foo<AnyT> where AnyT : any
{
    // Do we get a TypeLoadException for List<AnyT> when JIT compiling this method
    // or do we get an InvalidProgramException when the List<T> constructor is trying
    // to get an empty array of T for the backing store?
    static object Get() => new List<AnyT>();
}
```
I expect that this will be C# compile error. (I do not expect we will add `any` relaxation for `List<T>`.)

If you agree, the exact runtime behavior is not really interesting here. I can also just crash somewhere in the runtime like it is the case for invalid IL in general.

MichalStrehovsky replied:
> I expect that this will be C# compile error

I expect that too - this was pseudo-C# just to convey what I mean :).

> If you agree, the exact runtime behavior is not really interesting here.

The runtime behavior might be observable through `MakeGenericType`/`MakeGenericMethod` - if we allow `MakeGeneric` with byref-like types and someone uses it with a type where the `T` is not `any` and we don't block it, they'll get an oddly damaged type. If we do block it, the question is where to block it. The place where we validate constraints seems like a natural fit. But it will affect the above mentioned code paths too.

danmoseley commented:
Cc @steveharter

AaronRobinsonMSFT commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+    public abstract partial class Type
+    {
++        [Intrinsic]
++        public static bool IsByRefLike<T>();
```
That is true and something I did consider. The reason I didn't go down that path is because it is much harder to author a JIT intrinsic with that sort of statement, at least with my current knowledge of the JIT. I'll add the option though. I agree that if the general form of `typeof(T).<PROP>` were enabled we could improve several scenarios.

jkotas replied:
> it is much harder to author a JIT intrinsic with that sort of statement

It is not that much harder. (Search for `NI_System_Type_get_IsValueType` to see what it takes to have `typeof(T).IsValueType` treated as intrinsic.)

AaronRobinsonMSFT replied:
> It is not that much harder. (Search for NI_System_Type_get_IsValueType to see what it takes to have typeof(T).IsValueType treated as intrinsic.)

I misread @MichalStrehovsky's statement, he was making the same point. The fact that this path is already present makes me think it should be preferred. I'll remove the new API and go with making `typeof(T).IsByRefLike` an instrinsic.

RikkiGibson commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
+    public class GenericParameterSupportsAnyTypeAttribute : Attribute
```
I'm nervous about using the term "any type". For example, I'm not sure whether pointers would be permitted as type arguments as part of this design.

AaronRobinsonMSFT replied:
That is a good point. It also would seem to align with using the existing attribute flags mentioned above. I will reconsider this.

AaronRobinsonMSFT committed:
Review feedback

AaronRobinsonMSFT committed:
Remove words

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,75 @@
+
+## Proposal
+
+Support for the following would be indicated by the existing `RuntimeFeature.ByRefFields` mechanism.
```
I believe that we are going to finish `ByRefFields` in .NET 7. This is going to be preview feature at best in .NET 7. I think this feature needs to get a new flag under `RuntimeFeature`.

AaronRobinsonMSFT replied:
Are we certain the runtime side of this can't be finished in .NET 7? At present there do seem to be a fair number of changes, but I'm not seeing a specific red-flag that indicates this can't be completed. I agree that if something isn't in prior to the last preview before our first RC we aren't going to make.

jkotas replied:
We can declare this finished only after both runtime and Roslyn sides are done, the end-to-end have seen some use and the feedback has been incorporated. If we want to ship it before that, it should be marked as preview. We even have annotations to mark it as such now.

AaronRobinsonMSFT replied:
> We can declare this finished only after both runtime and Roslyn sides are done

That wasn't how I was defining this, but that does make it clear this won't be complete in .NET 7. What is the process for marking this in preview?

jkotas replied:
Design doc: https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md

This is what similar situation looked like in .NET 6 (snippet https://github.com/dotnet/runtime/blob/release/6.0/src/libraries/System.Runtime/ref/System.Runtime.cs):
```
    public static partial class RuntimeFeature
    {
        [System.Runtime.Versioning.RequiresPreviewFeaturesAttribute("Generic Math is in preview.", Url = "https://aka.ms/dotnet-warnings/generic-math-preview")]
        public const string VirtualStaticsInInterfaces = "VirtualStaticsInInterfaces";
    }
```

AaronRobinsonMSFT committed:
Update based on feedback

AaronRobinsonMSFT committed:
Add F# as being impacted.

AaronRobinsonMSFT commented:
@lambdageek for mono and @tgani-msft for C++/CLI FYI.

lambdageek commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,76 @@
+# Generics parameters of ByRefLike types
```
Just to clarify: this is specifically about `Span<Span<byte>>` not `Span<ref byte>`?

AaronRobinsonMSFT replied:
Correct. I believe that `Span<ref T>` would be slightly more complicated for all parties. This work is specifically about permitting `ref struct` types (ByRefLike) to be used as Generic parameters. Further details from the language side can be found at https://github.com/dotnet/csharplang/pull/5689.

jaredpar commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,76 @@
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; This IL sequence must resolve to a method implemented on `object`, or a default interface method.
+
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`.
```
Types like `Span<T>` have several APIs that conflict with the rules above. Most notable:

```c#
public Span(T[]? array);
public static implicit operator Span<T>(T[]? array);
```

How are we going to rationalize these APIs when we add `allow T: ref struct` to `Span<T>`?

RikkiGibson replied:
Jared and I have discussed this separately before. There are probably multiple viable ways to go but I think that exposing a compiler-known way to suppress constraint checks on specific APIs would do it. Then, I'm hoping that conversion errors at the use site will prevent the APIs from being used when the type arguments don't permit it.

```cs
using System;

namespace System.Diagnostics.CodeAnalysis
{
    public sealed class SuppressConstraintChecksAttribute : Attribute { }
}

ref struct Span<T> allow T : ref struct
{
    [SuppressConstraintChecks]
    public Span(T[]? array) { ... }

    [SuppressConstraintChecks]
    public static implicit operator Span<T>(T[]? array) { ... }
}

class Program
{
    static void M(Span<Span<int>> span)
    {
        var span1 = new Span<Span<int>>(null); // error: can't convert 'null' to type 'Span<int>[]'
        var span2 = (Span<Span<int>>)null; // error: can't convert 'null' to type 'Span<int>[]'
    }
}
```

This might leave a hole where the attribute could be used on methods which only violate type parameter constraints within the implementation, not in the signature. The user wouldn't be stopped from doing it at compile time but could get a runtime error. That might just be a case of "play stupid games, win stupid prizes".

Tagging @davidwrighton who had interest in a 'SuppressConstraintChecksAttribute' for a scenario with static interface members, IIRC.

AaronRobinsonMSFT replied:
> That might just be a case of "play stupid games, win stupid prizes".

That is how I am thinking of this feature. There are cases where the compiler can definitely determine if the user is doing something wrong, but also where it _might_ be doing something wrong. The [`Span<T>.ToArray()`](https://docs.microsoft.com/dotnet/api/system.span-1.toarray) is another example. If a user calls `ToArray()` and `T` is known to be ByRefLike, the compiler could issue a failure. However, if `T` isn't know at compile time then it would be up to the implementer to have handled this appropriately using something like the proposed intrinsi-fication of `typeof(T.).IsByRefLike` in this document. The updated `Span<T>.ToArray()` would look as follows and handle the case at runtime. The other APIs in question would have similar checks.

```diff
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public T[] ToArray()
{
+    if (typeof(T).IsByRefLike)
+        throw new InvalidProgramException();

    if (_length == 0)
        return Array.Empty<T>();

    var destination = new T[_length];
    Buffer.Memmove(ref MemoryMarshal.GetArrayDataReference(destination), ref _pointer.Value, (uint)_length);
    return destination;
}
```

AaronRobinsonMSFT replied:
It will be enforced but can all be detected at JIT time, which can then error once `T` is defined. The fact that `typeof(T).IsByRefLike` is a JIT intrinsic will permit it to detect that branch is always taken and the remaining instructions can be removed. In the event the user didn't put in that call the invalid instructions would then be processed but during the inspection of `T` it would be determined something is invalid and `InvalidProgramException` throw.

I suppose the example is bad since it is also throwing. Let me update it to reflect a more reasonable return.

AaronRobinsonMSFT replied:
Actually `ToArray` is a hard example to make something more reasonable. Imagine the return could be something that is used by the caller in a reasonable way. Unsure what that would be in this case though.

AaronRobinsonMSFT replied:
Thanks @RikkiGibson and @jaredpar. I'd not fully appreciated the compiler issues here.

jaredpar commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,76 @@
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; This IL sequence must resolve to a method implemented on `object`, or a default interface method.
```
The "must" here is throwing me off. My intuition is that `constrained.callvirt` would be an error when it resolved to `object` or a default interface method. Otherwise though it should be allowed.

AaronRobinsonMSFT replied:
That is a cumbersome description. Let me try again.

AaronRobinsonMSFT replied:
@jaredpar Let me know if this is clearer.

jaredpar commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,76 @@
+
+The `Reflection.Emit` API will need to be updated to respect the behavior of this flag. How it will handle support is an open question.
+
+## Open questions
```
There are other proposals out there to expand the set of allowed generic arguments. Most notably to include pointers. Do you anticipate this design will cause future compat issues with those proposals?

AaronRobinsonMSFT replied:
I can't see that causing an issue. We have a fair number of bits (10) remaining on `GenericParameterAttributes` after this one so we are still good to support both. I suppose we could look at devising a more extensive mechanism if there are future concerns.

AaronRobinsonMSFT committed:
Feedback and clarify wording for constrained calls.

AaronRobinsonMSFT commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,76 @@
+throw
+```
+
+When boxing due to a constrained call that cannot be made, instead of allocating a normal boxed object, an object of `InvalidBoxedObject` type will be created. It will have implementations of the various overridable object methods which throw `InvalidProgramException`, and interface dispatch shall have a special case for attempting to invoke an interface method on such an object, that will also t ...
````
@lambdageek This has changed to simply be a failure during the boxing attempt. I assume there is an easy spot to handle that boxing and fail.

jaredpar commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,83 @@
+
+- `throw` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `unbox` / `unbox.any` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `isinst` &ndash; Will always place `null` on stack.
```
Why would `isinst` always produce `null`? It would seem to be a completely valid pattern to do the following

```c#
void M<T>(T value) {
  if (value is Span<byte> span) {
  }
}
```

This pattern works for value types today.

AaronRobinsonMSFT replied:
I'm not sure how to make this work to be honest. I see the generated code is basically what you have below. However, a `box` instruction has language that indicates it will be boxed and that can't happen for ByRefLike types. I'd say this is invalid IL if `T` is ByRefLike and will fail during JIT.

I am interpreting this case from ECMA-335:

> If typeTok is a value type, the box instruction converts val to its boxed form. When typeTok is a 
non-nullable type (§I.8.2.4), this is done by creating a new object and copying the data from val 
into the newly allocated object. If it is a nullable type, this is done by inspecting val’s HasValue 
property; if it is false, a null reference is pushed onto the stack; otherwise, the result of boxing 
val’s Value property is pushed onto the stack. 

> If typeTok is a reference type, the box instruction does returns val unchanged as obj. 

> If typeTok is a generic parameter, the behavior of box instruction depends on the actual type at 
runtime. If this type is a value type it is boxed as above, if it is a reference type then val is not 
changed. However the type tracked by verification is always “boxed” typeTok for generic parameters, regardless of whether the actual type at runtime is a value or reference type. 

The statement "If this type is a value type it is boxed as above" is what I am focusing on.

AaronRobinsonMSFT replied:
I see. There is huge special case in the JIT that handles this case. It is commented as "match and import common box idioms". My description is valid and follows ECMA, but the JIT special cases this sequence and effectively elides the `box` instruction.

AaronRobinsonMSFT replied:
Basically, your concern is noted and that sequence should continue to work.

AaronRobinsonMSFT replied:
Recorded sequences and defined plan.

jaredpar commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,83 @@
+The following instructions are already set up to support this feature since their behavior will fail as currently defined due to the inability to box a ByRefLike type.
+
+- `throw` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `unbox` / `unbox.any` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
```
True they require an object reference but the following sequence is effectively how non-boxing casts work in IL: 

```il
        IL_000d: ldarg.0
        IL_000e: box !!T
        IL_0013: isinst [System.Private.CoreLib]System.Int32
        IL_0018: unbox.any [System.Private.CoreLib]System.Int32
        IL_001d: stloc.0
```

This doesn't actually put the `struct` on the heap hence my instinct is it would be valid for `ref struct`. If it's not valid then we effectively lose the ability to do pattern matching and type checks on `ref struct`.

AaronRobinsonMSFT replied:
The `box` instruction above doesn't actually box the argument?

AaronRobinsonMSFT replied:
See https://github.com/dotnet/runtime/pull/67129#discussion_r837976315

AaronRobinsonMSFT committed:
Add well-known IL sequences for boxing
Record APIs where compiler constraint analysis should be suppressed.
Defined analysis suppression attribute.

AaronRobinsonMSFT committed:
Minor nits.

jaredpar commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,123 @@
++   gpAcceptByRefLike = 0x0100 // type argument can be ByRefLike
+} CorGenericParamAttr;
+```
+
````
Consider an API currently defined as 

```c#
public void M<T>() { }
```

Given this design consider if it is changed to the following:

```c#
public void M<T>() allow T : ref struct { }
```

Does that represent a binary breaking change or not (assuming C# compiler emits the metadata as specified here)? Think it's overall desirable that this not be a binary breaking change as it allows existing APIs, that could be `ref struct` compatible, to be evolved to be such.

AaronRobinsonMSFT replied:
I was intending this to follow the same rules as if someone was relaxing an existing constraint. For example, if one converts

```csharp
public void M<T>() where T : unmanaged { }
```
to
```csharp
public void M<T>() { }
```

In my mind this is the same as being more permissible with now allowing `ref struct`. I'd argue the relaxing of constraints and expansion of permissible types should follow the same rules.

I've not checked if the current constraint behavior is a breaking change so perhaps someone can education me here, but that is what I would expect.

jkotas commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,123 @@
+throw
+```
+
+The `Reflection.Emit` API will be updated to respect ByRefLike support in Generics. The current Reflection API requires boxing of all types which makes usage of ByRefLike types impossible. Until the Reflection API can fully support ByRefLike types, this feature must be blocked when using Reflection. For example, API calls such as `MakeGenericType` / `MakeGenericMethod` will be invalid when `T` is ...
````
> The `Reflection.Emit` API will be updated to respect ByRefLike support in Generics.

What does it mean specifically? Does it require any new API or just fixing up some internal implementation details?

jkotas replied:
> For example, API calls such as `MakeGenericType` / `MakeGenericMethod` will be invalid when `T` is ByRefLike.

Why is that? I do not see anything fundamental that should prevent this. These APIs basically just call the type loader. If the type loader can create the type e.g. for the JIT, these APIs should be able to create the type as well.

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,123 @@
+
+## Special IL Sequences
+
+The following are IL sequences involving the `box` instruction. They are used in scenario optimizations and shall continue to be valid, even with ByRefLike types, in cases where the result can be computed at JIT time and elided safely. They will be added to the ECMA-335 addendum.
```
```suggestion
The following are IL sequences involving the `box` instruction. They are used for common C# language constructs and shall continue to be valid, even with ByRefLike types, in cases where the result can be computed at JIT time and elided safely. They will be added to the ECMA-335 addendum.
```

jkotas replied:
> in cases where the result can be computed at JIT time

What are those cases exactly? There are varying levels of optimizations and assumptions that the JIT can apply to these. The exact cases that are required to be have recognized by the JIT have to be enumerated as part of the spec. These cases will be required for functional correctness, they won't be an optional optimization anymore.

jkotas replied:
I think you probably want to say that the JIT is always required to recognize these sequences and never perform the box operation for them. 

Whether the JIT can compute the result as a constant or whether it calls a helper function to compute the result is JIT implementation detail.

jkotas commented on docs/design/features/byreflike-generics.md:
````diff
@@ -0,0 +1,123 @@
+}
+```
+
+Troublesome APIs:
````
I expect that there will be a lot more APIs like this that are troublesome. For example, `ArraySegment<T>` will need to allow T to be byref type (`static implicit operator ReadOnlySpan<T>(ArraySegment<T> segment)` would not compile otherwise) and APIs on `ArraySegment` that create arrays internally will be troublesome.

AaronRobinsonMSFT replied:
> would not compile otherwise) and APIs on ArraySegment that create arrays internally will be troublesome.

I don't see why this is the case. The proposed `SuppressConstraintChecksAttribute` attribute would let the compiler know validation should be suppressed for this and handled at run-time. What am I missing with how this attribute will work?

AaronRobinsonMSFT replied:
@jaredpar That is also how I understand this type too.  However, @jkotas mentioned the need for `ArraySegment<T>` to support ByRefLike types, which is the part I am not understanding. The `SuppressConstraintChecksAttribute` would make types that won't work in the context a run-time failure as opposed to the compiler handling the case. Or am I missing the distinction between `unsafe` and the flow of generic parameter constraints resolution?

jkotas replied:
I assumed that the compiler will always enforce the constraints, and `SuppressConstraintChecksAttribute` will be only used to mark methods that may fail runtime.

If `SuppressConstraintChecksAttribute` were to suppress constraint enforcement, it opens a whole new set of failure modes. For example, inspecting a method signature using reflection may throw new kinds of exceptions since the signature may contain malformed types due to suppressed constraints.

jkotas replied:
Specific example:
```
foreach (var m in typeof(Span<Span<int>>).GetMethods())
    foreach (var t in m.GetParameters())
        Console.WriteLine(t);
```
Does this throw an exception when it hits the operator that takes `ArraySegment`?

AaronRobinsonMSFT replied:
I would not expect that to fail since it is being used for Reflection purposes. This is similar to what I was assuming was your perspective regarding `MakeGenericType` / `MakeGenericMethod` above.

jkotas replied:
> I would not expect that to fail 

Reflection would need to return Type for `ArraySegment<Span<int>>` to make this succeed. Reflection only ever returns types with valid constraints. It means that `ArraySegment<T>` would need to allow T to be byref type.

AaronRobinsonMSFT replied:
Good point. Another option would be to skip over the types that aren't valid. This approach would be less than ideal but given the scope of this API it might not be too bad. The tradeoff here is between special casing the Reflection API or adding the ByRefLike support to types that can never be instantiated. In our offline conversation, you mentioned `abstract` types fall into this as well so I think the latter approach makes the most sense right now.

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,123 @@
+
+Troublesome APIs:
+- [`Span<T>`](https://docs.microsoft.com/dotnet/api/system.span-1)
+    - `public Span(T[]? array);`
```
The type loader rejects arrays of byref like types very early today (look for `IDS_CLASSLOAD_BYREFLIKEARRAY`). It is not possible to create a type for e.g. `Span<int>[]`. 

This restriction would prevent troublesome methods from being inspected using reflection when the type is instantiated over byref type. Attempt to get the method argument types would throw.

I think we will need to allow creating these types (e.g. allow `typeof(Span<int>[])`) and delay the failure till attempt to allocate the array allocation to make the reflection behavior reasonable.

AaronRobinsonMSFT replied:
I think we are going with the idea that using Reflection on the `Span<T>`/`ReadOnlySpan<T>` types may cause a `TypeLoadException` if `T` is ByRefLike.

AaronRobinsonMSFT committed:
Update docs/design/features/byreflike-generics.md

Co-authored-by: Jared Parsons <jaredpparsons@gmail.com>

AaronRobinsonMSFT commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,123 @@
+
+The expansion of metadata will impact at least the following:
+
+- ILDasm/ILAsm &ndash; https://github.com/dotnet/runtime
```
Agree. Will mention SRM and SRE here.

AaronRobinsonMSFT commented on docs/design/features/byreflike-generics.md:
```diff
@@ -0,0 +1,123 @@
+- `Span<TypedReference>` &ndash; Represents the general case where a ByRefLike type is used as a Generic parameter. This specific case would be desirable for a more efficient Reflection API.
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
```
That will be enforced by the TypeLoader with the `ref` field work already. I will call it out here though.

AaronRobinsonMSFT committed:
Review feedback

AaronRobinsonMSFT committed:
Update to design doc

AaronRobinsonMSFT committed:
Misspelling

AaronRobinsonMSFT commented:
Any other feedback on this design?

tgani-msft reviewed (commented):
Are these types importable via the COM metadata APIs? The C++ compiler uses only these.

AaronRobinsonMSFT commented:
As in, can they be acquired through the `IMetaDataImport` interface? Yes. A new enum value is being added though. That will need to be updated. Can you share an example of how the compiler uses `IMetaDataImport` so we can record the pattern?

AaronRobinsonMSFT committed:
Update byreflike-generics.md

tgani-msft commented:
> @tgani-msft As in, can they be acquired through the `IMetaDataImport` interface? Yes. A new enum value is being added though. That will need to be updated. Can you share an example of how the compiler uses `IMetaDataImport` so we can record the pattern?

I'll send you a link to the source code directly, via email.

RikkiGibson commented:
I think this came up elsewhere (maybe from @teo-tsirpanis), but is there any need to take special effort to make types "conditionally by-ref-like" depending on their type arguments? For example, imagine if we wanted to let people use nullable ref structs. It would be good to allow things like boxing, heap escape, etc. depending on the type arguments.

```cs
ref struct RS
{
    public ref int X;
}

namespace System
{
    // somehow we denote here that this is "only a ref struct if T is a ref struct".
    public struct Nullable<T> where T : struct allow T : ref struct // or whatever syntax we use
    {
        T value;
        // ...
    }
}

void M(RS? maybeRs, ref int X)
{
    if (maybeRs != null)
    {
        maybeRs.Value.X = ref X;
    }
    object obj = maybeRs; // error
}

void M2(int? maybeInt)
{
    object obj = maybeInt; // ok
}
```

It's possible the answer here is no, don't bother doing this, or punt it to something further down the line, but seems worth considering as part of this whole area.

AaronRobinsonMSFT commented:
> but is there any need to take special effort to make types "conditionally by-ref-like" depending on their type arguments?

That seems like a much broader feature given `T` would now influence the use of the Generic type. For example, if this were to be implemented we would also need to update the semantics of `IsByRefLikeAttribute` or create another mechanism that is predicated on other features. At present, I don't see anything here that would limit this, but do have some concerns about this feature and the complexity it would introduce in the type system and runtime in general.

AaronRobinsonMSFT commented:
The only prior art I can think of for this sort of thing is in Interop when a `T` influences the blittability of a type.

RikkiGibson commented:
FWIW, there are a few existing places in the compiler where the types of fields after substitution affect usage of a struct. For example: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzguATEBqAPmgJwFcBjBAAgGUAeAFQD4BYAKAG8XzPyABAZnNrkAZlACmAGwwBuFgF8WLbgCYeARgDsLdsy7kiMFBCGieAFnIBZVQAoasBPQBU5AA4ICASnKtys8gHp/cgB7AGsOLn1DYzNLJVtqYOAAK1EyJ1d3Lx8/QPJRAgJggjkgA==)

```cs
struct S<T>
{
    public T field;
}

class Program
{
    unsafe void M1(S<int>* ptr) { } // ok
    unsafe void M2(S<object>* ptr) { } // error
}
```

AaronRobinsonMSFT commented:
> FWIW, there are a few existing places in the compiler where the types of fields after substitution affect usage of a struct. For example: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzguATEBqAPmgJwFcBjBAAgGUAeAFQD4BYAKAG8XzPyABAZnNrkAZlACmAGwwBuFgF8WLbgCYeARgDsLdsy7kiMFBCGieAFnIBZVQAoasBPQBU5AA4ICASnKtys8gHp/cgB7AGsOLn1DYzNLJVtqYOAAK1EyJ1d3Lx8/QPJRAgJggjkgA==). (it sounds like this might be very similar/related to blittability.)

Agreed, but not metadata. For example, the following might be unintuitive from a reflection standpoint:

```csharp
struct S<T> allow T : ref struct
            ref struct when T : ref struct
{ }

Console.WriteLine(typeof(S<>).IsByRefLike); // False
Console.WriteLine(typeof(S<int>).IsByRefLike); // False
Console.WriteLine(typeof(S<Span<int>>).IsByRefLike);  // True
```

hez2010 commented:
How about enabling overload by generic constraints so we can have both `Foo<T> where T : struct` and `Foo<T> where T : ref struct`?

AaronRobinsonMSFT commented:
>@AaronRobinsonMSFT, can this be merged?

I think so. It is accurate regarding the current state of support in the runtime.

> Presumably we're fine updating the design doc as the design evolves and we don't need to block the PR on having 100% of the answers set in stone?

Agreed.
